### ISSUE TITLE
syslog-ng compiles but fails to start on macOS Big Sur

### ISSUE THREAD

**Issue by czanik** [syslog-ng/syslog-ng#3385]
It is possible to compile syslog-ng on the latest macOS beta. But when trying to start it, I get an error message:
```
bash-3.2# /usr/local/sbin/syslog-ng -Fvde
iv_fd_register: called with invalid fd 6 (maxfd=-1)
Abort trap: 6
bash-3.2#
```
Based on the above error message the problem might be in ivykis. I use ivykis from brew (which works perfectly on the latest stable macOS).
```
Peters-MacBook-Pro:~ czanik$ brew info ivykis
ivykis: stable 0.42.4 (bottled)
Async I/O-assisting library
https://sourceforge.net/projects/libivykis/
/usr/local/Cellar/ivykis/0.42.4 (88 files, 230.0KB) *
  Poured from bottle on 2020-08-10 at 16:06:25
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/ivykis.rb
License: LGPL-2.1
==> Dependencies
Build: autoconf ✔, automake ✔, libtool ✔
==> Analytics
install: 22 (30 days), 47 (90 days), 142 (365 days)
install-on-request: 23 (30 days), 48 (90 days), 143 (365 days)
build-error: 0 (30 days)
```

**Comment by czanik** [comment]
Deleted the ivykis brew package. Compiled syslog-ng again with bundled ivykis, but the problem remained the same.

**Comment by bazsi** [comment]
Hi,

Ccing Lennart, who might have an idea about the problem.

**Comment by Kokan** [comment]
@czanik does brew install run ivykis unit tests ? are those passing ?

**Comment by czanik** [comment]
@Kokan Huh. It's not that simple. ```brew install``` just downloads a binary package. That was in theory built with ```make check``` and https://formulae.brew.sh/formula/ivykis does not list any build errors. But now, when I try to build it from source, ```brew``` complained, that I need a more recent Xcode installed. Right now I have the latest from the AppStore, but that's not good enough, so I'm downloading the latest beta from a developers website...

**Comment by czanik** [comment]
No luck. Even with updated Xcode the error remains the same.

**Comment by michaelklishin** [comment]
@czanik you need to symlink the SDK, see the discussion in https://github.com/Homebrew/brew/issues/7803.

**Comment by buytenh** [comment]
I'm guessing that the current `RLIMIT_NOFILE` limit is `RLIM_INFINITY` on your platform when ivykis starts, and that that causes `maxfd` to be `-1`.

As a quick hack, does this work?

```diff --git a/src/iv_fd.c b/src/iv_fd.c
index c31ad58..6b3e725 100644
--- a/src/iv_fd.c
+++ b/src/iv_fd.c
@@ -35,6 +35,12 @@ static void sanitise_nofile_rlimit(int euid)
        struct rlimit lim;
 
        getrlimit(RLIMIT_NOFILE, &lim);
+
+       if (lim.rlim_cur == RLIM_INFINITY) {
+               maxfd = 131072;
+               return;
+       }
+
        maxfd = lim.rlim_cur;
 
        if (euid) {
```

**Comment by czanik** [comment]
@buytenh thanks for looking at it

@MrAnno do you still have a working development environment on the Beta macOS? One of the beta updates broke Xcode and I never managed to compile syslog-ng ever since, even after reinstalling both Xcode and Homebrew...

**Comment by czanik** [comment]
OK. Reinstalled my MacBook from scratch, as going through all the MacOS beta releases did not help :) 

Added the patch from @buytenh and it helped. Now I get different errors (sorry for not quoting properly, but I have a half working keyboard with an unknown layout...):

bash-3.2$ syslog-ng/syslog-ng -V
syslog-ng 3 (3.30.1)
Config version: 3.29
Installer-Version: 3.30.1
Revision: 
Compile-Date: Dec  2 2020 16:44:54
Module-Directory: /usr/local/lib/syslog-ng
Module-Path: /usr/local/lib/syslog-ng
Include-Path: /usr/local/share/syslog-ng/include
Error opening plugin module; module='json-plugin', error='dlopen(/usr/local/lib/syslog-ng/libjson-plugin.dylib, 9): Library not loaded: /usr/local/lib/libjson-c.3.dylib\x0a  Referenced from: /usr/local/lib/syslog-ng/libjson-plugin.dylib\x0a  Reason: image not found'
Bus error: 10

bash-3.2$ syslog-ng/syslog-ng -F
iv_tls_user_register: called after iv_init
Abort trap: 6

**Comment by bazsi** [comment]
This is already a syslog-ng related problem, you probably have multiple
versions of syslog-ng in the same prefix.

**Comment by czanik** [comment]
No. It's a fresh macOS install and I only tried to compile syslog-ng 3.30.1 on it ever.

The json problem was resolved by installing json-c on the system instead of using the bundled json-c. But both commands still fail:

```
bash-3.2$ /usr/local/sbin/syslog-ng -V
syslog-ng 3 (3.30.1)
Config version: 3.29
Installer-Version: 3.30.1
Revision: 
Compile-Date: Dec  2 2020 21:57:49
Module-Directory: /usr/local/lib/syslog-ng
Module-Path: /usr/local/lib/syslog-ng
Include-Path: /usr/local/share/syslog-ng/include
Bus error: 10
bash-3.2$ /usr/local/sbin/syslog-ng -F
iv_tls_user_register: called after iv_init
Abort trap: 6
bash-3.2$ 
```

**Comment by bazsi** [comment]
I usually see this error when we do a version change, and we there's a
module in our plugin directory that links to a different version of
libsyslog-ng.so, thereby pulling it in twice.

Does that ring a bell?

Perhaps remove the plugins from the installation path and run make install
again.

**Comment by czanik** [comment]
No, it's not the case here. It was a first installation on a freshly installed system. But I removed everything anyway and had the same results after installation.

If you or anyone else from the team wants access to my MacBook for debugging we can most likely arrange that (if my Internet provider did not change that), just e-mail me at my company address. After the fresh install I don't have anything personal (yet) on that box.

**Comment by MrAnno** [comment]
@buytenh Wow, you are absolutely right. Thanks.

This is triggered by syslog-ng, because we do the following:
https://github.com/syslog-ng/syslog-ng/blob/2d594f664a5d8c01081c96c966ac3d05ef8675d4/lib/gprocess.c#L685-L692


After this call, `getrlimit(RLIMIT_NOFILE, &lim)` returns `lim.rlim_cur == 9223372036854775807` (`RLIM_INFINITY`) which will be negative one if it's assigned to a signed int (`maxfd`), hence the abort.

Note: macOS Big Sur's man says
```
setrlimit() now returns with errno set to EINVAL in places that historically succeeded.
It no longer accepts "rlim_cur = RLIM_INFINITY" for RLIM_NOFILE.  Use "rlim_cur = min(OPEN_MAX, rlim_max)".
```

**Comment by MrAnno** [comment]
The syslog-ng part of the fix can be found in .

We could have used `sysconf()` to get OPEN_MAX in a portable manner, something like this:
a portable mannel
 but I think

**Comment by MrAnno** [comment]
The syslog-ng part of the fix can be found in #3522.

We could have used `sysconf()` to get `OPEN_MAX` in a portable manner, something like this:
https://github.com/syslog-ng/syslog-ng/blob/013df12e397eb3bfc40d00f0c1b7cb228d28483f/modules/affile/directory-monitor.c#L57-L67

but I think it's not necessary in this case, as `rlim_cur` can be set to `rlim_max` on all platforms except for macOS.

@buytenh #3522 fixes the `iv_fd_register()` abort in syslog-ng, but I think ivykis might need to prepare for `rlim_cur == RLIM_INFINITY` as well.

**Comment by rcmelendez** [comment]
Thanks for the quick fix guys! It works now, just tested on Big Sur 11.1 and syslog-ng 3.30.1.
